app/cluster/state/etcd.go

```diff
@@ -0,0 +1,279 @@
+	envConfigOnce    sync.Once
+)
+
+const (
+	modeRequestKeyPattern        = `/%s/server/%s/mode`       // /<releaseName>/server/<serverIndex>/mode
+	workspacesRequestsKeyPattern = `/%s/server/%s/workspaces` // /<releaseName>/server/<serverIndex>/workspaces
+)
+
+var _ cluster.ModeProvider = &ETCDManager{}
+
+type ETCDConfig struct {
+	ReleaseName          string
+	ServerIndex          string
+	Endpoints            []string
+	dialKeepAliveTime    time.Duration
+	dialKeepAliveTimeout time.Duration
+	etcdWatchTimeout     time.Duration
+	etcdGetTimeout       time.Duration
+	dialTimeout          time.Duration
+}
+
+type modeRequestValue struct {
+	Mode   servermode.Mode `json:"mode"`
+	AckKey string          `json:"ack_key"`
+}
+
+type modeAckValue struct {
+	Status servermode.Mode `json:"status"`
+}
+
+type workspacesRequestsValue struct {
+	Workspaces string `json:"workspaces"` // comma separated workspaces
+	AckKey     string `json:"ack_key"`
+}
+
+type workspacesAckValue struct {
+	Status string `json:"status"`
+}
+
+func EnvETCDConfig() *ETCDConfig {
+	endpoints := strings.Split(config.GetEnv("ETCD_HOSTS", "127.0.0.1:2379"), `,`)
+	releaseName := config.GetEnv("RELEASE_NAME", `multitenantv1`)
+	serverIndex := config.GetInstanceID()
+
+	envConfigOnce.Do(func() {
+		config.RegisterDurationConfigVariable(time.Duration(15), &etcdGetTimeout, true, time.Second, "etcd.getTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(3), &etcdWatchTimeout, true, time.Second, "etcd.watchTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(30), &keepaliveTime, true, time.Second, "etcd.keepaliveTime")
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, true, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, true, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		ReleaseName:          releaseName,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+func (manager *ETCDManager) Ping() error {
```

Comment:
This will ping only once so the name is misleading. One would expect something over the wire each time this method is called but that's not the case. Perhaps consider renaming (e.g. `HadConnected`) or consider making it ping for real.

Reply:
Good, point. I wanted to do a health check. Fixed.

Reply:
Did you push the fix? :thinking: